benchmarks/src/main/java/org/elasticsearch/benchmark/common/util/IntArrayBenchmark.java

```diff
@@ -0,0 +1,98 @@
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OperationsPerInvocation;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+@Warmup(iterations = 5)
+@Measurement(iterations = 7)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@State(Scope.Thread)
+@Fork(1)
+@OperationsPerInvocation(2621441) // NOCOMMIT s/41/40/ when fixed aligned read
+public class IntArrayBenchmark {
+    static final int SIZE = IntArrayBenchmark.class.getAnnotation(OperationsPerInvocation.class).value();
+
+    @Param({ "array", "paged_bytes_array", "composite_256kb", "composite_262344b", "composite_1mb" })
+    private String type;
+
+    private IntArray read;
+
+    @Setup
+    public void init() throws IOException {
+        IntArray ints = BigArrays.NON_RECYCLING_INSTANCE.newIntArray(SIZE);
+        for (int i = 0; i < SIZE; i++) {
+            ints.set(i, i);
+        }
+        BytesStreamOutput out = new BytesStreamOutput();
+        ints.writeTo(out);
+        read = IntArray.readFrom(new ReleasableBytesReference(bytesImpl(out.bytes()), () -> {}).streamInput());
+    }
+
+    private BytesReference bytesImpl(BytesReference bytes) {
+        if (type.equals("array")) {
+            return new BytesArray(bytes.toBytesRef());
+        }
+        if (type.equals("paged_bytes_array")) {
+            if (bytes instanceof PagedBytesReference == false) {
+                throw new AssertionError("expected PagedBytesReference but saw [" + bytes.getClass() + "]");
+            }
+            return bytes;
+        }
+        if (type.startsWith("composite_")) {
+            int size = Math.toIntExact(ByteSizeValue.parseBytesSizeValue(type.substring("composite_".length()), "type").getBytes());
+            List<BytesReference> references = new ArrayList<>();
+            for (int from = 0; from < bytes.length(); from += size) {
+                int sliceSize = Math.min(size, bytes.length() - from);
+                references.add(new BytesArray(bytes.slice(from, Math.min(from + size, sliceSize)).toBytesRef()));
+            }
+            BytesReference ref = CompositeBytesReference.of(references.toArray(BytesReference[]::new));
+            if (ref instanceof CompositeBytesReference == false) {
+                throw new AssertionError("expected CompositeBytesReference but saw [" + bytes.getClass() + "]");
+            }
+            return ref;
+        }
+        throw new IllegalArgumentException("unsupported [type] " + type);
+    }
+
+    @Benchmark
+    public long read() {
+        int res = 0;
+        for (int i = 0; i < SIZE; i++) {
+            res = res ^ read.get(i);
```

Comment:
NIT: I think it might give a better benchmark to store `this.read` into a local variable before the loop. I don't think the JIT is able to remove the field access here, hard to tell how much that matters here though.

Reply:
I can give it a go and see. I thought it'd be able to avoid both the things you mentioned but I dunno! :scientist:

Reply:
It doesn't seem to make a difference. This morning the fast path clocks in at .22ns either way. My CPU claims 2.3GHz which implies .43ns per cycle. So when this is an array we're looking at about half a cycle per read. Now I'm curious.

Reply:
Here's the hot block:
```
         ╭     0x00007f0eb8ff0106:   jmpq   0x00007f0eb8ff018b
         │     0x00007f0eb8ff010b:   nopl   0x0(%rax,%rax,1)
  0.00%  │↗    0x00007f0eb8ff0110:   mov    %ebx,%r10d                   ;*iload_1 {reexecute=0 rethrow=0 return_oop=0}
         ││                                                              ; - org.elasticsearch.benchmark.common.util.IntArrayBenchmark::read@11 (line 95)
  1.74%  ││ ↗  0x00007f0eb8ff0113:   mov    %r10d,%ebx
         ││ │  0x00007f0eb8ff0116:   shl    $0x2,%ebx
  0.00%  ││ │  0x00007f0eb8ff0119:   movslq %ebx,%rbx
  0.00%  ││ │  0x00007f0eb8ff011c:   xor    0x10(%rsi,%rbx,1),%r11d
  2.31%  ││ │  0x00007f0eb8ff0121:   xor    0x14(%rsi,%rbx,1),%r11d
  0.00%  ││ │  0x00007f0eb8ff0126:   xor    0x18(%rsi,%rbx,1),%r11d
  1.13%  ││ │  0x00007f0eb8ff012b:   xor    0x1c(%rsi,%rbx,1),%r11d
  1.72%  ││ │  0x00007f0eb8ff0130:   xor    0x20(%rsi,%rbx,1),%r11d
  1.70%  ││ │  0x00007f0eb8ff0135:   xor    0x24(%rsi,%rbx,1),%r11d
  1.68%  ││ │  0x00007f0eb8ff013a:   xor    0x28(%rsi,%rbx,1),%r11d
  1.72%  ││ │  0x00007f0eb8ff013f:   xor    0x2c(%rsi,%rbx,1),%r11d
  1.64%  ││ │  0x00007f0eb8ff0144:   xor    0x30(%rsi,%rbx,1),%r11d
  1.76%  ││ │  0x00007f0eb8ff0149:   xor    0x34(%rsi,%rbx,1),%r11d
  1.71%  ││ │  0x00007f0eb8ff014e:   xor    0x38(%rsi,%rbx,1),%r11d
  3.52%  ││ │  0x00007f0eb8ff0153:   xor    0x3c(%rsi,%rbx,1),%r11d
  1.74%  ││ │  0x00007f0eb8ff0158:   xor    0x40(%rsi,%rbx,1),%r11d
  1.72%  ││ │  0x00007f0eb8ff015d:   xor    0x44(%rsi,%rbx,1),%r11d
  1.70%  ││ │  0x00007f0eb8ff0162:   xor    0x48(%rsi,%rbx,1),%r11d
  1.70%  ││ │  0x00007f0eb8ff0167:   xor    0x4c(%rsi,%rbx,1),%r11d      ;*ixor {reexecute=0 rethrow=0 return_oop=0}
         ││ │                                                            ; - org.elasticsearch.benchmark.common.util.IntArrayBenchmark::read@23 (line 95)
  1.64%  ││ │  0x00007f0eb8ff016c:   mov    %r10d,%ebx
  0.00%  ││ │  0x00007f0eb8ff016f:   add    $0x10,%ebx                   ;*iinc {reexecute=0 rethrow=0 return_oop=0}
         ││ │                                                            ; - org.elasticsearch.benchmark.common.util.IntArrayBenchmark::read@25 (line 94)
  0.00%  ││ │  0x00007f0eb8ff0172:   cmp    %edx,%ebx
         │╰ │  0x00007f0eb8ff0174:   jl     0x00007f0eb8ff0110           ;*if_icmpge {reexecute=0 rethrow=0 return_oop=0}
         │  │                                                            ; - org.elasticsearch.benchmark.common.util.IntArrayBenchmark::read@8 (line 94)
```

This reads to me like it's inlined the array access and unrolled the loop. That's with the field read.

Reply:
Right, the field lookup is irrelevant here then. I guess the unexpected performance is probably due to op fusion in the CPU. Seems possible to fuse those xors maybe in some form, that's a little beyond my area of expertise :)